flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/JoinTestPrograms.java

```diff
@@ -0,0 +1,274 @@
+
+        INNER_JOIN_WITH_DUPLICATE_KEY =
+                TableTestProgram.of(
+                                "inner-join-with-duplicate-key", "inner join with duplicate key")
+                        .setupTableSource(SOURCE_A)
+                        .setupTableSource(SOURCE_B)
+                        .setupTableSink(
+                                SinkTestStep.newBuilder("MySink")
+                                        .addSchema("a1 int", "b1 int")
+                                        .consumedBeforeRestore(Row.of(2, 2))
+                                        .consumedAfterRestore(new Row[]{})
+                                        .build())
+                        .runSql(
+                                "insert into MySink "
+                                        + "SELECT a1, b1 FROM A JOIN B ON a1 = b1 AND a1 = b3")
+                        .build();
+
+        INNER_JOIN_WITH_NON_EQUI_JOIN =
+                TableTestProgram.of(
+                                "inner-join-with-non-equi-join", "inner join with non-equi join")
+                        .setupTableSource(SOURCE_A)
+                        .setupTableSource(SOURCE_B)
+                        .setupTableSink(
+                                SinkTestStep.newBuilder("MySink")
+                                        .addSchema("a1 int", "b1 int")
+                                        .consumedBeforeRestore(Row.of(3, 3))
+                                        .consumedAfterRestore(new Row[]{})
+                                        .build())
+                        .runSql(
+                                "insert into MySink "
+                                        + "SELECT a1, b1 FROM A JOIN B ON a1 = b1 AND a2 > b2")
+                        .build();
+
+        String query1 = "SELECT SUM(a2) AS a2, a1 FROM A GROUP BY a1";
+        String query2 = "SELECT SUM(b2) AS b2, b1 FROM B GROUP BY b1";
+
+        INNER_JOIN_WITH_EQUAL_PK =
+                TableTestProgram.of("inner-join-with-equal-pk", "inner join with equal pk")
+                        .setupTableSource(SOURCE_A)
+                        .setupTableSource(SOURCE_B)
+                        .setupTableSink(
+                                SinkTestStep.newBuilder("MySink")
+                                        .addSchema("a1 int", "b1 int")
+                                        .consumedBeforeRestore(
+                                                Row.of(1, 1), Row.of(2, 2), Row.of(3, 3))
+                                        .consumedAfterRestore(new Row[]{})
+                                        .build())
+                        .runSql(
+                                String.format(
+                                        "INSERT INTO MySink SELECT a1, b1 FROM (%s) JOIN (%s) ON a1 = b1",
+                                        query1,
+                                        query2))
+                        .build();
+
+        INNER_JOIN_WITH_PK =
```

Comment:
@dawidwys While regenerating the plans multiple times, I noticed that the output changes for this test could either be:

```[+I[1, 1], +I[3, 2], +I[2, 2], -U[3, 2], -U[2, 2], +I[1, 3]]``` 
Or
```[+I[1, 1], +I[1, 3]]```

The change logs are equivalent once one applies the `-U`s.  Thoughts on this kind of situation?